### PR TITLE
9 encoder

### PR DESCRIPTION
**Pull request by KwiatkowskiML**
This pull request introduces a comprehensive implementation of a Transformer model, including its components such as the encoder, decoder, and various supporting layers. It also defines constants for model configuration in a separate file for better modularity. The most important changes are grouped into the following themes:

### New Constants for Model Configuration
* Added constants in `constants.py` for model parameters such as `D_MODEL_SIZE`, `NUMBER_OF_LAYERS`, `ATTENTION_HEADS_NUMBER`, `DROPOUT`, and `HIDDEN_LAYER_SIZE` to centralize configuration values.

#### Core Transformer Components
* Implemented the `Transformer` class, which integrates the encoder, decoder, embedding layers, positional encodings, and projection layer. This class provides methods for encoding, decoding, and projecting outputs.
* Introduced `Encoder` and `Decoder` classes, each composed of multiple layers (`EncoderBlock` and `DecoderBlock`) with residual connections and normalization.
* Added the `build_transformer` function to construct a complete Transformer model with configurable parameters such as vocabulary size, sequence length, and model dimensions.

#### Supporting Layers and Enhancements
* Added `ResidualConnection`, `ProjectionLayer`, and updated `LayerNormalization` to include type annotations for improved clarity and functionality. [[1]](diffhunk://#diff-fada037ad086638e65c7ae77e3d223963e9afaa26326aab0ea718f4013176e43L98-R263) [[2]](diffhunk://#diff-fada037ad086638e65c7ae77e3d223963e9afaa26326aab0ea718f4013176e43L38-R50)
* Updated the `MultiHeadAttention` class to include a `forward` method for computing attention with proper reshaping and projections.

#### Documentation and Code Readability
* Added inline comments to clarify tensor transformations and expected shapes in various methods, such as in `InputEmbeddings` and `MultiHeadAttention`. [[1]](diffhunk://#diff-fada037ad086638e65c7ae77e3d223963e9afaa26326aab0ea718f4013176e43R17) [[2]](diffhunk://#diff-fada037ad086638e65c7ae77e3d223963e9afaa26326aab0ea718f4013176e43R90-R106)

These changes collectively provide a robust implementation of a Transformer model, following best practices for modularity, readability, and extensibility.

### CODE REVIEW

File: constants.py
Comment (KKKUBAKKK) [inline constants.py:1]: Dobry plan z tym plikiem

File: model.py [hunk 1]
```diff
@@ -95,19 +118,146 @@ def attention(query, key, value, mask, dropout: nn.Dropout):
...
+
+
+class ProjectionLayer(nn.Module):
+    def __init__(self, d_model: int, vocab_size: int) -> None:
+        super().__init__()
+        self.proj = nn.Linear(d_model, vocab_size)
+
+    def forward(self, x):
+        # (Batch, Seq_len, d_model) --> (Batch, Seq_len, Vocab_size)
+        return torch.log_softmax(self.proj(x), dim=-1)
```
Comment (KKKUBAKKK) [inline model.py:190]: Tutaj w kodzie z repo z filmiku jest:         return self.proj(x)
Ale zakładam, że może jeszcze później to zmienimy?

File: model.py [hunk 2]
```diff
@@ -95,19 +118,146 @@ def attention(query, key, value, mask, dropout: nn.Dropout):
...
+        self.residual_connections = nn.ModuleList([ResidualConnection(dropout) for _ in range(2)])
+
+    def forward(self, x, src_mask):
+        x = self.residual_connections[0](x, lambda x: self.self_attention_block(x, x, x, src_mask))
+        x = self.residual_connections[1](x, self.feed_forward_block)
+        return x
+
+
+class Encoder(nn.Module):
+    def __init__(self, layers: nn.ModuleList):
```
Comment (KKKUBAKKK) [inline model.py:146]: W przykładowym kodzie jest jeszcze features: int, jako argument

File: model.py [hunk 3]
```diff
@@ -95,19 +118,146 @@ def attention(query, key, value, mask, dropout: nn.Dropout):
...
+        self.residual_connections = nn.ModuleList([ResidualConnection(dropout) for _ in range(3)])
+
+    def forward(self, x, encoder_output, src_mask, tgt_mask):
+        x = self.residual_connections[0](x, lambda x: self.self_attention_block(x, x, x, tgt_mask))
+        x = self.residual_connections[1](x, lambda x: self.cross_attention_block(x, encoder_output, encoder_output, src_mask))
+        x = self.residual_connections[2](x, self.feed_forward_block)
+        return x
+
+class Decoder(nn.Module):
+    def __init__(self, layers: nn.ModuleList):
```
Comment (KKKUBAKKK) [inline model.py:172]: W przykładowym kodzie jest jeszcze features: int, jako argument

File: model.py [hunk 4]
```diff
@@ -95,19 +118,146 @@ def attention(query, key, value, mask, dropout: nn.Dropout):
...
+    decoder_blocks = []
+    for _ in range(layers_num):
+        decoder_self_attention_block = MultiHeadAttention(d_model, heads_number, dropout)
+        decoder_cross_attention_block = MultiHeadAttention(d_model, heads_number, dropout)
+        feed_forward_block = FeedForwardBlock(d_model, d_ff, dropout)
+        decoder_block = DecoderBlock(decoder_self_attention_block, decoder_cross_attention_block, feed_forward_block, dropout)
+        decoder_blocks.append(decoder_block)
+
+    # Create the encoder and decoder
+    encoder = Encoder(nn.ModuleList(encoder_blocks))
```
Comment (KKKUBAKKK) [inline model.py:249]: Jeden argument mniej w konstruktorze

File: model.py [hunk 5]
```diff
@@ -95,19 +118,146 @@ def attention(query, key, value, mask, dropout: nn.Dropout):
...
+    decoder = Decoder(nn.ModuleList(decoder_blocks))
```
Comment (KKKUBAKKK) [inline model.py:250]: Jeden argument mniej w konstruktorze

**Review by KKKUBAKKK — APPROVED** [review]
